client/txn.go

```diff
@@ -0,0 +1,135 @@
+package client
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+)
+
+type Txn interface {
+	// If takes a list of comparison. If all comparisons passed in succeed,
+	// the operations passed into Then() will be executed. Or the operations
+	// passed into Else() will be executed.
+	If(cs ...Cmp) Txn
+
+	// Then takes a list of operations. The Ops list will be executed, if the
+	// comparisons passed in If() succeed.
+	Then(ops ...Op) Txn
+
+	// Else takes a list of operations. The Ops list will be executed, if the
+	// comparisons passed in If() fail.
+	Else(ops ...Op) Txn
+
+	// Commit tries to commit the transaction.
+	Commit() (*TxnResponse, error)
+}
+
+type txn struct {
+	// Name of the kv client, which will be used in CURP propose id generation
+	name string
+	// The client running the CURP protocol, communicate with all servers.
+	curpClient curpClient
+	// The auth token
+	token string
+
+	mu    sync.Mutex
+	cif   bool
+	cthen bool
+	celse bool
+
+	cmps []*xlineapi.Compare
+	sus  []*xlineapi.RequestOp
+	fas  []*xlineapi.RequestOp
+}
+
+func (txn *txn) If(cs ...Cmp) Txn {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	if txn.cif {
+		panic("cannot call If twice!")
+	}
+	if txn.cthen {
+		panic("cannot call If after Then!")
+	}
+	if txn.celse {
+		panic("cannot call If after Else!")
+	}
+
+	txn.cif = true
+
+	for i := range cs {
+		txn.cmps = append(txn.cmps, (*xlineapi.Compare)(&cs[i]))
+	}
+
+	return txn
+}
+
+func (txn *txn) Then(ops ...Op) Txn {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	if txn.cthen {
+		panic("cannot call Then twice!")
+	}
+	if txn.celse {
+		panic("cannot call Then after Else!")
+	}
+
+	txn.cthen = true
+
+	for _, op := range ops {
+		txn.sus = append(txn.sus, op.toRequestOp())
+	}
+
+	return txn
+}
+
+func (txn *txn) Else(ops ...Op) Txn {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	if txn.celse {
+		panic("cannot call Else twice!")
+	}
+
+	txn.celse = true
+
+	for _, op := range ops {
+		txn.fas = append(txn.fas, op.toRequestOp())
+	}
+
+	return txn
+}
+
+func (txn *txn) Commit() (*TxnResponse, error) {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	var krs []*xlineapi.KeyRange
+	for _, cmp := range txn.cmps {
+		kr := &xlineapi.KeyRange{Key: cmp.Key, RangeEnd: cmp.RangeEnd}
+		krs = append(krs, kr)
+	}
+	req := &xlineapi.TxnRequest{Compare: txn.cmps, Success: txn.sus, Failure: txn.fas}
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &txn.token,
+		RequestWrapper: &xlineapi.RequestWithToken_TxnRequest{
+			TxnRequest: req,
+		},
+	}
+	proposeId := txn.generateProposeId()
+	cmd := xlineapi.Command{Keys: krs, Request: &requestWithToken, ProposeId: proposeId}
+	res, err := txn.curpClient.propose(&cmd, false)
+	if err != nil {
+		return nil, err
+	}
+	return (*TxnResponse)(res.CommandResp.GetTxnResponse()), err
+}
+
+// Generate a new `ProposeId`
+func (txn *txn) generateProposeId() string {
```

Comment:
Please add a TODO to update the propose id. FYI: https://github.com/xline-kv/Xline/blob/84c685ac4b311ec035076b295e192c65644f85b9/curp-external-api/src/cmd.rs#L84

Reply:
I've updated it as you recommended.